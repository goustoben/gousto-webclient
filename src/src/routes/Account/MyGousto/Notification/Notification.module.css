@value ScreenSMMin from 'styles/screenSizes.module.css';
@value avenirHeavyFontStack from 'styles/typography.module.css';

.notificationLink {
  text-decoration: none;
  color: black;
  cursor: pointer;

  &:hover {
    color: black;
  }
}

.notificationTitle {
  font-weight: bold;
  font-size: 15px;
  font-family: avenirHeavyFontStack;
}

.notificationMessage {
  display: block;

  @media ScreenSMMin {
    display: inline;
  }
}

.notificationLink span {
  text-decoration: none;
}