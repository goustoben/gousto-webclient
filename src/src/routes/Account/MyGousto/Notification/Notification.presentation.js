import React from 'react'
import PropTypes from 'prop-types'
import { Alert } from 'goustouicomponents'
import css from './Notification.css'

const propTypes = {
  type: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  message: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
}

const NotificationPresentation = ({ type, title, message, url }) => (
  <div>
    <Alert type={type}>
      <a href={url} className={css.notificationLink}>
        <span className={css.notificationTitle}>{title} </span>
        <span className={css.notificationMessage}>{message}</span>
      </a>
    </Alert>
  </div>
)

NotificationPresentation.propTypes = propTypes

export { NotificationPresentation }
