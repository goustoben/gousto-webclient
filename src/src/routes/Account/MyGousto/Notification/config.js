const notificationConfig = ({
  expired: {
    message: 'Your card is expired. Please update now.',
    type: 'danger',
    title: 'Payment Method',
    url: 'my-details',
  },
  toExpire: {
    message: 'Your card is about to expire soon. Please update now.',
    type: 'danger',
    title: 'Payment Method',
    url: 'my-details',
  },
  amendDelivery: {
    message: 'We’ve moved your delivery to the next available delivery day.',
    type: 'info',
    title: 'Delivery changes',
    url: 'my-deliveries',
  },
  selectOrder: {
    message: 'Otherwise our chef will send their best selection.',
    type: 'info',
    title: 'Choose your recipes for your upcoming order before 12pm.',
    url: 'menu',
  },
  referAFriend: {
    message: 'for every friend you refer this weekend. Invite friends now.',
    type: 'info',
    title: 'Get £30 credit',
    url: 'my-referrals',
    linkTrackingType: 'refer_friend',
    startDate: '2019-11-22',
    endDate: '2019-11-24',
  },
  sustainabilityPledge: {
    message: 'And this is just the beginning. Find out more.',
    type: 'info',
    title: 'We’ve solved our 50% pledge.',
    url: 'https://www.gousto.co.uk/unwrapped',
    linkTrackingType: 'sustainability_pledge',
    startDate: '2020-01-10',
    endDate: '2020-02-29',
  },
})

export { notificationConfig as config }

