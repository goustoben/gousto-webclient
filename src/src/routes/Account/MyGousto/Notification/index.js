export { NotificationLogic as Notification } from './Notification.logic'
