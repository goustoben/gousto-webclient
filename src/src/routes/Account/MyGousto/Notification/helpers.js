import moment from 'moment'

import { config } from './config'

export const checkCardExpiryDate = (card, now) => {
  let bannerToShow
  const expiryDate = moment(card.get('expiryDate'))

  if (!card.size) return

  if (now.isBefore(expiryDate) && expiryDate.diff(now, 'days') <= 30) {
    bannerToShow = 'toExpire'
  } else if (now.isSameOrAfter(expiryDate)) {
    bannerToShow = 'expired'
  }

  return bannerToShow
}

export const checkAmendedDeliveryDate = (orders) => {
  const alternateDeliveryDays = orders.filter(order => order.get('state') === 'pending' && order.get('originalDeliveryDay')).toArray()
  if (alternateDeliveryDays.length > 0) {
    return 'amendDelivery'
  }
}

export const checkOrderAwaitingSelection = (orders, now) => {
  const notifications = orders
    .filter(order => order.get('state') === 'pending' && order.get('default') === '1')
    .filter(order => moment(order.get('whenCutoff')).isSame(now, 'day'))
    .toArray()

  if (notifications.length >= 1) {
    return 'selectOrder'
  }
}

export const checkRafOffer = (now) => {
  const { startDate, endDate } = config.referAFriend
  const rafExpiry = moment(now).isBetween(startDate, endDate)

  if (rafExpiry) {
    return 'referAFriend'
  }
}

export const checkSustainabilityPledge = (now) => {
  const { startDate, endDate } = config.sustainabilityPledge
  const showBanner = moment(now).isBetween(startDate, endDate)

  return showBanner ? 'sustainabilityPledge' : undefined
}

export const priority = (type = '') => {
  switch(type) {
  case 'danger':
    return 3
  case 'warning':
    return 2
  case 'notify':
    return 1
  default:
    return 0
  }
}

export const sortNotifications = (a, b) => (
  priority(b.type) - priority(a.type)
)
