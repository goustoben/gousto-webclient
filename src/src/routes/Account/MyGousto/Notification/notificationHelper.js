import moment from 'moment'
import { config } from './config'

export const checkCardExpiryDate = (card, now) => {
  let bannerToShow
  const expiryDate = moment(card.get('expiryDate'))

  if (!card.size) return

  if (now.isBefore(expiryDate) && expiryDate.diff(now, 'days') <= 30) {
    bannerToShow = 'toExpire'
  } else if (now.isSameOrAfter(expiryDate)) {
    bannerToShow = 'expired'
  }

  return bannerToShow
}

export const checkAmendedDeliveryDate = (orders) => {
  const alternateDeliveryDays = orders.filter(order => order.state === 'pending' && order.original_delivery_day).toArray()
  if (alternateDeliveryDays.length > 0) {
    return 'amendDelivery'
  }
}

export const checkOrderAwaitingSelection = (orders, now) => {
  const notifications = orders
    .filter(order => order.state === 'pending' && order.default === '1')
    .filter(order => moment(order.when_cutoff).isSame(now, 'day'))
    .toArray()

  if (notifications.length >= 1) {
    return 'selectOrder'
  }
}

export const checkRafOffer = (now) => {
  const { startDate, endDate } = config.referAFriend
  const rafExpiry = moment(now).isBetween(startDate, endDate)

  if (rafExpiry) {
    return 'referAFriend'
  }
}
