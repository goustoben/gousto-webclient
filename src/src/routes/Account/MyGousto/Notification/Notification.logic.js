import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Immutable from 'immutable'
import moment from 'moment'
import { NotificationPresentation } from './Notification.presentation'
import { config } from './config'
import { checkCardExpiryDate, checkAmendedDeliveryDate, checkOrderAwaitingSelection, checkRafOffer } from './notificationHelper'

class NotificationLogic extends Component {

  static propTypes = {
    card: PropTypes.instanceOf(Immutable.Map),
    orders: PropTypes.instanceOf(Immutable.Map),
  }

  static defaultProps = {
    card: Immutable.Map({}),
    orders: Immutable.Map({}),
  }

  state = {
    bannersToShow: [],
  }

  componentDidMount() {
    const { card, orders } = this.props
    const now = moment()

    const banners = [
      checkCardExpiryDate(card, now),
      checkAmendedDeliveryDate(orders),
      checkOrderAwaitingSelection(orders, now),
      checkRafOffer(now)
    ].filter(banner => banner)

    this.setState({
      bannersToShow: banners
    })
  }

  sortBanners = (a, b) => {
    if (b.type === 'danger') return 1
    if (b.type === 'warning' && a.type !== 'danger') return 1
    if (b.type === 'notify' && a.type !== 'danger' && a.type !== 'warning') return 1

    return 0
  }

  render() {
    const { bannersToShow } = this.state

    const notificationBannerDetails = bannersToShow.map(banner => (
      {
        message: config[banner].message,
        type: config[banner].type,
        title: config[banner].title,
        url: config[banner].url,
      }
    ))
      .sort((a, b) => this.sortBanners(a, b))

    return (
      <div>
        {
          notificationBannerDetails.map((banner, index) => {
            return index < 2 ?
              < NotificationPresentation key={banner.title} message={banner.message} type={banner.type} title={banner.title} url={banner.url} />
              : null
          })
        }
      </div>
    )

  }
}

export { NotificationLogic }
