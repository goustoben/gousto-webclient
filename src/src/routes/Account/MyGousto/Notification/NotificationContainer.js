import { connect } from 'react-redux'
import { trackNotificationLinkClick } from '../actions/tracking'
import { NotificationLogic } from './Notification.logic'

export const NotificationContainer = connect(null, {
  trackNotificationLinkClick
})(NotificationLogic)
