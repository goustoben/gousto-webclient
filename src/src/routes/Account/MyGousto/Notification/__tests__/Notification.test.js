import React from 'react'
import moment from 'moment'
import { mount } from 'enzyme'
import Immutable from 'immutable'

import { config } from '../config'
import {
  checkCardExpiryDate,
  checkAmendedDeliveryDate,
  checkOrderAwaitingSelection,
  checkRafOffer,
  checkSustainabilityPledge,
} from '../helpers'

import { NotificationLogic as Notification } from '../Notification.logic'

config.referAFriend.startDate = '2019-01-01'
config.referAFriend.endDate = '2019-01-01'

describe('Notification component', () => {
  let wrapper
  let card
  let orders
  let now

  describe('checkCardExpiryDate', () => {
    beforeEach(() => {
      now = moment('2019-08-15')
    })

    it('should return "Expired" if card expiry date is in current month or earlier', () => {
      card = Immutable.Map({
        lastFourDigits: "1234",
        expiryDate: '2019-08',
      })

      const result = checkCardExpiryDate(card, now)
      expect(result).toEqual('expired')
    })

    it('should return "Expiring" if card expiry date is within one month of now', () => {
      card = Immutable.Map({
        lastFourDigits: "1234",
        expiryDate: '2019-09',
      })

      const result = checkCardExpiryDate(card, now)
      expect(result).toEqual('toExpire')
    })

    it('should return undefined if no cards are expired or expiring', () => {
      card = Immutable.Map({
        lastFourDigits: "1234",
        expiryDate: '2019-11',
      })

      const result = checkCardExpiryDate(card, now)
      expect(result).toEqual(undefined)
    })

    it('should return undefined if there are no card details', () => {
      card = Immutable.Map({})

      const result = checkCardExpiryDate(card, now)
      expect(result).toEqual(undefined)
    })
  })

  describe('checkAmendedDeliveryDate', () => {
    it('should return "Amend Delivery" if delivery date has been amended', () => {
      orders = Immutable.fromJS({
        1234: {
          state: 'pending',
          originalDeliveryDay: true,
        }
      })

      const result = checkAmendedDeliveryDate(orders)
      expect(result).toEqual('amendDelivery')
    })

    it('should return undefined if no delivery dates have been amended', () => {
      orders = Immutable.fromJS({
        1234: {
          state: 'pending',
          originalDeliveryDay: false,
        }
      })

      const result = checkAmendedDeliveryDate(orders)
      expect(result).toEqual(undefined)
    })

    it('should return undefined if no orders', () => {
      orders = Immutable.Map({})

      const result = checkAmendedDeliveryDate(orders)
      expect(result).toEqual(undefined)
    })
  })

  describe('checkOrderAwaitingSelection', () => {
    beforeEach(() => {
      now = moment('2019-08-15')
    })

    it('should return "Select order" if upcoming order has not been selected', () => {
      orders = Immutable.fromJS({
        1234: {
          state: 'pending',
          default: '1',
          whenCutoff: moment('2019-08-15 11:59:59')
        }
      })

      const result = checkOrderAwaitingSelection(orders, now)
      expect(result).toEqual('selectOrder')
    })

    it('should return undefined if no upcoming orders are unselected', () => {
      orders = Immutable.fromJS({
        1234: {
          state: 'pending',
          default: '1',
          whenCutoff: moment('2019-08-23 11:59:59')
        }
      })

      const result = checkOrderAwaitingSelection(orders, now)
      expect(result).toEqual(undefined)
    })

    it('should return undefined if there are no orders', () => {
      orders = Immutable.Map({})

      const result = checkOrderAwaitingSelection(orders, now)
      expect(result).toEqual(undefined)
    })
  })

  describe('checkRafOffer', () => {
    beforeEach(() => {
      now = moment('2019-08-15')

      config.referAFriend.startDate = '2019-01-01'
      config.referAFriend.endDate = '3019-01-01'
    })

    afterEach(() => {
      config.referAFriend.startDate = '2019-01-01'
      config.referAFriend.endDate = '2019-01-01'
    })

    it('should show "Refer a friend" notification if current date is between start and end date', () => {
      const result = checkRafOffer(now)
      expect(result).toEqual('referAFriend')
    })

    it('should return undefined if current date is not between start and end date', () => {
      config.referAFriend.endDate = '2019-01-01'

      const result = checkRafOffer(now)
      expect(result).toEqual(undefined)
    })
  })

  describe('checkSustainabilityPledge', () => {
    beforeEach(() => {
      now = moment('2020-01-15')

      config.sustainabilityPledge.startDate = '2020-01-01'
      config.sustainabilityPledge.endDate = '3020-01-01'
    })

    it('should show the sustainability pledge notification if current date is between start and end date', () => {
      const result = checkSustainabilityPledge(now)
      expect(result).toEqual('sustainabilityPledge')
    })

    it('should return undefined if current date is not between start and end date', () => {
      config.sustainabilityPledge.endDate = '2019-01-01'

      const result = checkSustainabilityPledge(now)
      expect(result).toEqual(undefined)
    })
  })

  describe('notifications', () => {
    beforeEach(() => {
      const currentMonth = moment().format('YYYY-MM')
      config.referAFriend.startDate = '2019-01-01'
      config.referAFriend.endDate = '3019-01-01'
      card = Immutable.Map({
        lastFourDigits: "1234",
        expiryDate: currentMonth,
      })
      orders = Immutable.fromJS({
        1234: {
          state: 'committed',
          originalDeliveryDay: true,
        }
      })
    })

    it('should call tracking action on click if notification has linkTrackingType', () => {
      const mockTrackNotificationLinkClick = jest.fn()
      window.location.assign = jest.fn()
      wrapper = mount(<Notification card={card} orders={orders} trackNotificationLinkClick={mockTrackNotificationLinkClick} />)

      wrapper.find(`[href="my-referrals"]`).simulate('click')
      expect(mockTrackNotificationLinkClick).toHaveBeenCalledWith(config.referAFriend.linkTrackingType)
      expect(window.location.assign).toHaveBeenCalledWith(config.referAFriend.url)
    })
  })
})
