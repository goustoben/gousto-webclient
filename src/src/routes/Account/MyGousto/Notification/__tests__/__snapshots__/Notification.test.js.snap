// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification component notifications should order the notifications by type in the following order danger > warning > notify 1`] = `
Array [
  Object {
    "message": "Your card is expired. Please update now.",
    "title": "Payment Method",
    "type": "danger",
    "url": "my-details",
  },
  Object {
    "message": "Weâ€™ve moved your delivery to the next available delivery day.",
    "title": "Delivery changes",
    "type": "info",
    "url": "my-deliveries",
  },
]
`;
